src/work/SimpleDEX.sol

```diff
@@ -1,223 +1,253 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.0;
 
-import "./IERC20.sol";
-import "@openzeppelin/contracts/utils/math/SafeMath.sol";
-
-/**
- * @title Simple Decentralized Exchange (DEX)
- * @dev A simplified version of a DEX allowing users to swap between two tokens (TokenA and TokenB).
- */
-contract SimpleDEX {
-    using SafeMath for uint256;
-
-    address public owner;
-    IERC20 public tokenA;
-    IERC20 public tokenB;
-    uint256 public rateTokenAtoTokenB = 100; // Example rate: 1 TokenA = 100 TokenB
-    uint256 public feePercentage; // Fee percentage, e.g., 1 means 1%
+// Import the IERC20 interface and SafeERC20 library from OpenZeppelin
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+// Import the ReentrancyGuard contract from OpenZeppelin for reentrancy protection
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+// Import the Ownable contract from OpenZeppelin for access control
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+// SimpleDEX contract inherits from ReentrancyGuard and Ownable contracts
+contract SimpleDEX is ReentrancyGuard, Ownable {
+    // Use SafeERC20 for safe token transfers and approvals
+    using SafeERC20 for IERC20;
+
+    // Immutable variables for token addresses to save gas on storage reads
+    IERC20 public immutable tokenA;
+    IERC20 public immutable tokenB;
+
+    // Exchange rate from tokenA to tokenB
+    uint256 public rateTokenAtoTokenB;
+    // Fee percentage for token swaps (in basis points, where 10000 = 100%)
+    uint256 public feePercentage;
+    // Flag to indicate if the contract is paused
     bool public isPaused;
-
-    bool private locked;
-
+    // Flag to indicate if the emergency stop is activated
+    bool public emergencyStop;
+    // Slippage tolerance for token swaps (in basis points, where 10000 = 100%)
+    uint256 public slippageTolerance;
+    // Minimum swap amount allowed
+    uint256 public minSwapAmount;
+    // Maximum swap amount allowed
+    uint256 public maxSwapAmount;
+
+    // Mapping to store whitelisted tokens
+    mapping(address => bool) public whitelistedTokens;
```

Comment:
doesnt need to be public, have a method where an addr is passed in and return bool on if its whitelisted or not